File: app/api/schemas/data_maintenance.py
from __future__ import annotations

from typing import Any, Dict, List, Optional

from pydantic import BaseModel


class DataSourceStatusOut(BaseModel):
    source_id: str
    label: str
    path: str
    kind: str
    exists: bool
    status: str
    file_size_bytes: Optional[int] = None
    mtime: Optional[str] = None
    row_count: Optional[int] = None
    asset_count: Optional[int] = None
    start_date: Optional[str] = None
    end_date: Optional[str] = None
    freshness_days: Optional[int] = None
    days_since_latest: Optional[int] = None
    is_blocking: bool = False
    freshness_reason: Optional[str] = None
    notes: List[str] = []
    child_count: Optional[int] = None
    sample_children: Optional[List[str]] = None
    calendar_count: Optional[int] = None
    instrument_counts: Optional[Dict[str, int]] = None
    feature_dir_count: Optional[int] = None


class DataPathAuditOut(BaseModel):
    generated_at: str
    overall_status: str
    blocking_status: str = "OK"
    blockers: List[Dict[str, Any]] = []
    recommendations: List[Dict[str, Any]] = []
    status_counts: Dict[str, int]
    sources: List[DataSourceStatusOut]


class RunDailyDataMaintenanceIn(BaseModel):
    dry_run: bool = False
    refresh_factor_registry: bool = True
    refresh_stock_screen: bool = True
    run_radar_smoke: bool = True
    run_external_updater: bool = False
    updater_id: Optional[str] = None


class DataMaintenanceRunOut(BaseModel):
    run_id: str
    generated_at: str
    dry_run: bool
    overall_status: str
    audit: Dict[str, Any]
    audit_before: Dict[str, Any]
    steps: List[Dict[str, Any]]
    artifacts: Optional[Dict[str, str]] = None

File: web/src/lib/api/data-maintenance.ts
import { fetchJson } from "@/lib/api/client";

export type DataSourceStatus = {
  source_id: string;
  label: string;
  path: string;
  kind: string;
  exists: boolean;
  status: string;
  file_size_bytes?: number | null;
  mtime?: string | null;
  row_count?: number | null;
  asset_count?: number | null;
  start_date?: string | null;
  end_date?: string | null;
  freshness_days?: number | null;
  days_since_latest?: number | null;
  is_blocking?: boolean;
  freshness_reason?: string | null;
  notes?: string[];
  child_count?: number | null;
  sample_children?: string[] | null;
  calendar_count?: number | null;
  instrument_counts?: Record<string, number> | null;
  feature_dir_count?: number | null;
  database_backed?: boolean;
  package_version?: string | null;
  available?: Record<string, boolean> | null;
  latest_query?: Record<string, unknown> | null;
};

export type DataHealthItem = {
  source_id: string;
  label?: string;
  status?: string;
  path?: string;
  reason?: string;
  updater_id?: string | null;
};

export type DataRecommendation = {
  source_id: string;
  status?: string;
  updater_id: string;
  label?: string;
  reason?: string;
};

export type DataPathAudit = {
  generated_at: string;
  overall_status: string;
  blocking_status?: "OK" | "WARN" | "BLOCKED";
  blockers?: DataHealthItem[];
  recommendations?: DataRecommendation[];
  status_counts: Record<string, number>;
  sources: DataSourceStatus[];
};

export type RunDailyMaintenancePayload = {
  dry_run?: boolean;
  refresh_factor_registry?: boolean;
  refresh_stock_screen?: boolean;
  run_radar_smoke?: boolean;
  run_external_updater?: boolean;
  updater_id?: string | null;
};

export type DataMaintenanceRun = {
  run_id: string;
  generated_at: string;
  dry_run: boolean;
  overall_status: string;
  audit: DataPathAudit;
  audit_before: DataPathAudit;
  steps: Record<string, unknown>[];
  artifacts?: Record<string, string> | null;
};

export async function getDataPathAudit(): Promise<DataPathAudit> {
  return fetchJson<DataPathAudit>("/api/data-maintenance/paths");
}

export async function runDailyDataMaintenance(payload: RunDailyMaintenancePayload): Promise<DataMaintenanceRun> {
  return fetchJson<DataMaintenanceRun>("/api/data-maintenance/daily-update", {
    method: "POST",
    body: JSON.stringify(payload),
  });
}

export async function getLatestDataMaintenanceReport(): Promise<DataMaintenanceRun> {
  return fetchJson<DataMaintenanceRun>("/api/data-maintenance/latest");
}
